tests/beta/tapes/test_tape.py

```diff
@@ -0,0 +1,667 @@
+        spy.assert_called_once()
+
+
+class TestParameters:
+    """Tests for parameter processing, setting, and manipulation"""
+
+    @pytest.fixture
+    def make_tape(self):
+        params = [0.432, 0.123, 0.546, 0.32, 0.76]
+
+        with QuantumTape() as tape:
+            qml.RX(params[0], wires=0)
+            qml.Rot(*params[1:4], wires=0)
+            qml.CNOT(wires=[0, "a"])
+            qml.RX(params[4], wires=4)
+            expval(qml.PauliX(wires="a"))
+            probs(wires=[0, "a"])
+
+        return tape, params
+
+    def test_parameter_processing(self, make_tape):
+        """Test that parameters are correctly counted and processed"""
+        tape, params = make_tape
+        assert tape.num_params == len(params)
+        assert tape.trainable_params == set(range(len(params)))
+        assert tape.get_parameters() == params
+
+    def test_set_trainable_params(self, make_tape):
+        """Test that setting trainable parameters works as expected"""
+        tape, params = make_tape
+        trainable = {0, 2, 3}
+        tape.trainable_params = trainable
+        assert tape._trainable_params == trainable
+        assert tape.num_params == 3
+        assert tape.get_parameters() == [params[i] for i in tape.trainable_params]
+
+        # add an additiona trainable parameter
+        trainable = {1, 2, 3, 4}
+        tape.trainable_params = trainable
+        assert tape._trainable_params == trainable
+        assert tape.num_params == 4
+        assert tape.get_parameters() == [params[i] for i in tape.trainable_params]
+
+    def test_changing_params(self, make_tape):
+        """Test that changing trainable parameters works as expected"""
+        tape, params = make_tape
+        trainable = {0, 2, 3}
+        tape.trainable_params = trainable
+        assert tape._trainable_params == trainable
+        assert tape.num_params == 3
+        assert tape.get_parameters() == [params[i] for i in tape.trainable_params]
+        assert tape.get_parameters(free_only=False) == params
+
+    def test_set_trainable_params_error(self, make_tape):
+        """Test that exceptions are raised if incorrect parameters
+        are set as trainable"""
+        tape, _ = make_tape
+
+        with pytest.raises(ValueError, match="must be positive integers"):
+            tape.trainable_params = {-1, 0}
+
+        with pytest.raises(ValueError, match="must be positive integers"):
+            tape.trainable_params = {0.5}
+
+        with pytest.raises(ValueError, match="has at most 5 parameters"):
+            tape.trainable_params = {0, 7}
+
+    def test_setting_parameters(self, make_tape):
+        """Test that parameters are correctly modified after construction"""
+        tape, params = make_tape
+        new_params = [0.6543, -0.654, 0, 0.3, 0.6]
+
+        tape.set_parameters(new_params)
+
+        for pinfo, pval in zip(tape._par_info.values(), new_params):
+            assert pinfo["op"].data[pinfo["p_idx"]] == pval
```

Comment:
:grimacing:

Reply:
lol. I was wondering whether to turn `pinfo` into a namedtuple? That way you could do
```python
assert pinfo.op.data[pinfo.p_idx] == pval
```

But, I'm cognizant about not making `p_info` turn back into the behemoth that `variable_deps` was. I just need it to store enough information about the parameters for tape execution/gradients